### ISSUE TITLE
Show Main Window On Second Run - Drop Error Msg

### ISSUE THREAD

**Issue by war59312** [OpenVPN/openvpn-gui#114]
Hi,

Currently if you attempt to run a second copy of OpenVPN GUI, you get:

> OpenVPN GUI is already running.

Would be a lot nicer if instead of the error. The OpenVPN GUI simply reapears on screen.

Thanks,

Will

**Comment by selvanair** [comment]
We have been asked for something like this in the past, but its yet not clear to me what exactly is the desired behaviour. The GUI is a tray application docks on to the tray, so there is no window to show. One could show the status window if any connections are active but that could be irritating too. And there is nothing to show if no connections are active. If the second instance is stated with --connect some-config, we could try to start it, but that's rarely the case.

**Comment by war59312** [comment]
It's for situations to help out newbies (my grandmother) who don't realize OpenVPN was already running.

Anyhow.. Here is how I envision it.

So with OpenVPN just sitting in the system tray (not connected), on the second run it should attempt to connect using the last supplied config file and try to connect with last supplied settings. If no config has been previously used, then simply ask the user which config to use. If no configs exist, then warn user as such.

If OpenVPN is already connected while sitting in the system tray, on the second run it should open the status window, but only the first time. Show a warning to the user about how it's already running and that in the future a second run will do nothing. So no annorying error and no annoyng status window. Nothing.

If a second instance is started with any kid of command line arguments, then they should take effect immediately. Maters not if connected already or not. Override it. As in, disconnect and reconnect using the command line arguments given.

Hope that helps?

PS: Just a temp solution. I intend to use a transparent proxy via the router (OpenWRT). But still I think it's better than just the error.

**Comment by chipitsine** [comment]
no-no-no, "grandmother case" is very good opportunity to improve user experience. if openvpn is good enough for grandmothers, it is good for everybody.

let us do our best to help grandmothers in using openvpn!

**Comment by mattock** [comment]
@chipitsine : :smile: yes, I see what you mean. That said, in this particular case the best user experience is one you don't think about (i.e. background service).

**Comment by chipitsine** [comment]
for this particular case I can suggest to use something like https://www.join.me/ , teamviewer or other "desktop sharing" software.

also, I suggest to pay some attention to "openvpn-gui is already running" case. there might be openvpn-gui from other users, should we set some "per user" ipc thing to distinguish between different users openvpn-gui running ?

and, the best will be the most simple, so stopping complaining about "already running" will be the best solution for now. and, if there will be ongoing issues, we can pay more attention to it

**Comment by Deantwo** [comment]
As already mentioned above, there is no "main window" to show. And showing the connection status window would make very little since.

But what you could do is show a balloon-tip from the tray icon.
![image](https://cloud.githubusercontent.com/assets/2676134/21680386/b7ed4d62-d349-11e6-83b4-dd92d2a01b87.png)
This should also help with locating the OpenVPN menu in case it is hidden.

The problem with this is that in Window 8/10 the balloon-tips have been replaced with the new (and slightly more annoying) toasts notifications.
![image](https://cloud.githubusercontent.com/assets/2676134/21680551/b89b2a76-d34a-11e6-9ad4-800c41c44b14.png)

**Comment by selvanair** [comment]
Hi,

On Thu, Jan 5, 2017 at 8:16 AM, Ilya Shipitsin <notifications@github.com>
wrote:

> such an introductory baloon might be very promising way of interacting
> with user.
>

I'm also leaning towards a balloon tip + a flashing tray icon (?) to point
the user to the running instance. . It will be useful in the first instance
too although Windows will most likely hide the tray icon the first time...

As for multiple users running the GUI in parallel its already supported in
the sense that you can switch user and start another instance. The old GUI
itself will not be active as the session is switched but any connections
started by it will continue in the background.

Selva

>

**Comment by war59312** [comment]
Decided to switch her over to a Router running OpenWRT, along with transparent proxy, and be done with it. :)

Yes, a Balloon tip sounds like a good idea to me.

**Comment by larsen0815** [comment]
I have another case where the current message is disturbing. I want to start my VPN-connection by using a shell command like this:
````
"C:\Program Files\OpenVPN\bin\openvpn-gui.exe" --connect companyvpn.ovpn
````
This works for the first time, but when I then disconnect and don't exit the GUI, I cannot use that command anymore as said message pops up. Would be a lot more useful for me if the connection was started again.

+1 for balloon tip
+1 for not showing anything when "--connect" is used

**Comment by Deantwo** [comment]
@larsen0815, it might be a separate issue altogether. But maybe when you started OpenVPN GUI with a `--connect` commandline argument like that, it should exit the GUI when you disconnect. Or maybe a new commandline argument should be added to do this; for example `--exitafterdisconnect` or something.

**Comment by larsen0815** [comment]
@Deantwo Yes, you may be right. I was aware of this before, that even if the message popup was gone, it doesn't guarantee that "--connect" will work. If it will, there is no need for a "--exitafterdisconnect". If it won't, that would be a separate issue.

I don't use really use the GUI, but I have only found a way to open a connection by using "openvpn-gui.exe"

**Comment by selvanair** [comment]
I've a feature branch that does this (https://github.com/selvanair/openvpn-gui/tree/already-running)
(i) Show a balloon if already running
(ii) send commands such as connect, disconnect, status, exit etc. to a running instance

Eg., with the GUI running do
openvpn-gui.exe --command connect <profile-name>  (omit .ovpn from profile name)
(see commit message for more details)
: 
It has been left idle for long and needs some love to finish it.. Feedback welcome.

**Comment by larsen0815** [comment]
I have seen the build instructions, but I'm afraid I don't have the time to compile that branch (as I can live with the current behaviour, even if I think it could be improved).

That being said, if you would be so nice to provide a Windows binary, I could test it in a production scenario.

**Comment by selvanair** [comment]
A test binary is here (64 bit only)
https://github.com/selvanair/openvpn-gui/releases/tag/v11.4.0.6_AR

**Comment by larsen0815** [comment]
Looks very good already!
Some issues documented in https://github.com/selvanair/openvpn-gui/issues/5

**Comment by selvanair** [comment]
Handled in PR #188

**Comment by war59312** [comment]
Thanks everyone!
